### PR TITLE
Various fixes for running real experiments

### PR DESCRIPTION
**Pull request by MatthewWilkes**
## Description
This adds log messages for tasks around creating a participant
for easier debugging, fixes a case where log messages fail
if the session hasn't begun yet and reorders createParticipant
to prevent the session being restarted, clearing the log messages.

## Motivation and Context
Without this, when running with the MTurk recruiter there is something like race condition which causes either the message not to be enqueued and the transaction to fail, or for the list of queued messages to be dropped.

## How Has This Been Tested?
This was developed by adding the `ps:exec` buildpack and using PDB directly on a heroku server running in sandbox mode.

### CODE REVIEW

**Comment by MatthewWilkes** [comment]
FYI, @suchow and @jessesnyder - this is needed to run live experiments.

File: dallinger/experiment_server/experiment_server.py [hunk 1]
```diff
@@ -747,21 +747,22 @@ def create_participant(worker_id, hit_id, assignment_id, mode):
     # Ping back to the recruiter that one of their participants has joined:
     recruiters.for_experiment(exp).notify_recruited(participant)
```
Comment (jessesnyder) [inline dallinger/experiment_server/experiment_server.py]: Tiny thing, but worth it I think, to reduce the mental load:
```
+    recruiter = recruiters.for_experiment(exp)

     # Ping back to the recruiter that one of their participants has joined:
-    recruiters.for_experiment(exp).notify_recruited(participant)
+    recruiter.notify_recruited(participant)

     overrecruited = exp.is_overrecruited(waiting_count)
     if not overrecruited:
         # We either had no quorum or we have not overrecruited, inform the
         # recruiter that this participant will be seeing the experiment
-        recruiters.for_experiment(exp).notify_using(participant)
+        recruiter.notify_using(participant)
```

File: dallinger/db.py [hunk 2]
```diff
@@ -134,16 +134,24 @@ def wrapper(*args, **kw):
 # Reset outbox when session begins
 @event.listens_for(Session, 'after_begin')
 def after_begin(session, transaction, connection):
+    logger.debug(
+            'Clearing message queue due to begin: {}'.format(session.info))
```
Comment (jessesnyder) [inline dallinger/db.py]: Playing in debug mode just now, I see that this is being called about 10x/second. Not sure what's triggering this(?)
```
14:59:46 web.1      |  2018-03-26 14:59:46,038 Clearing message queue due to begin: {'outbox': []}
14:59:46 web.1      |  2018-03-26 14:59:46,086 Clearing message queue due to begin: {'outbox': []}
14:59:46 web.1      |  2018-03-26 14:59:46,234 Clearing message queue due to begin: {'outbox': []}
14:59:46 web.1      |  2018-03-26 14:59:46,288 Clearing message queue due to begin: {'outbox': []}
14:59:46 web.1      |  2018-03-26 14:59:46,356 Clearing message queue due to begin: {'outbox': []}
14:59:46 web.1      |  2018-03-26 14:59:46,419 Clearing message queue due to begin: {'outbox': []}
14:59:46 web.1      |  2018-03-26 14:59:46,473 Clearing message queue due to begin: {'outbox': []}
14:59:46 web.1      |  2018-03-26 14:59:46,565 Clearing message queue due to begin: {'outbox': []}
14:59:46 web.1      |  2018-03-26 14:59:46,671 Clearing message queue due to begin: {'outbox': []}
14:59:46 web.1      |  2018-03-26 14:59:46,741 Clearing message queue due to begin: {'outbox': []}
14:59:46 web.1      |  2018-03-26 14:59:46,817 Clearing message queue due to begin: {'outbox': []}
14:59:46 web.1      |  2018-03-26 14:59:46,880 Clearing message queue due to begin: {'outbox': []}
14:59:46 web.1      |  2018-03-26 14:59:46,925 Clearing message queue due to begin: {'outbox': []}
14:59:46 web.1      |  2018-03-26 14:59:46,987 Clearing message queue due to begin: {'outbox': []}
```

File: dallinger/experiment_server/experiment_server.py (hunk 1 above)
Comment (MatthewWilkes) [reply dallinger/experiment_server/experiment_server.py]: Good call

File: dallinger/db.py (hunk 2 above)
Comment (MatthewWilkes) [reply dallinger/db.py]: Yeah, I got this too, it's much more than I expected. I mean, it's the right use of the debug log level in that it's useful for debugging only, but we use debug quite a lot in real experiments.

**Comment by MatthewWilkes** [comment]
@jessesnyder The issue was that the implementation of `notify_using` for `MTurkRecruiter` needs to look at attributes of the `Participant` model, but as we've just flushed the DB the participant could be stale. I'm not sure exactly why this happens, as I'd only expect this on commit, but we are running this from within a `SET TRANSACTION ISOLATION LEVEL SERIALIZABLE` block, so maybe that causes it to not survive a flush. Anyway, when we try and access an attribute of participant it causes a SQL query to be issued, which as the first one since the `flush()` causes a transaction to begin. This causes `session.info['outbox']` to be set to an empty list, overwriting the message we just wrote.

The effect of all this is that the `quorum` websocket never has any messages on it except for ping, so only the person who causes an experiment to reach quorum will enter it (as they get a full response from createParticipant), the others will be stuck on a loading bar that never receives updates.

**Pull request closed without merging**